File: ios-native/ClubFuoco/Features/ClubDetail/EventDetailSheet.swift
import SwiftUI

/// Everything we hold about one event, opened by tapping its card on a club
/// page. Port of "Club Fuoco Event Page" from the design project.
///
/// The design was drawn from measured coverage (docs/event-detail-data-brief.md)
/// and covers four real states, all of which this must survive:
///   1. rich      — flyer, copy, several linked DJs, age, capacity
///   2. no lineup — 29% of events bill no artists at all
///   3. floor     — no flyer AND no copy: title, time, venue, promoter only
///   4. unlinked  — credits that are real names but not DJs we hold
///
/// Nothing here is faked when absent: a missing flyer removes the image, a
/// missing lineup says so, and the two fields that lie (`cost`, `venue_capacity`)
/// are gated behind validators on ClubEvent rather than shown raw.
struct EventDetailSheet<DJPage: View>: View {
    let event: ClubEvent
    /// Credits resolved to DJ pages — by RA id first, name as legacy fallback.
    let djFor: (LineupCredit) -> FeaturedDJ?
    /// The DJ page for a lineup row. It is presented BY this sheet rather than
    /// handed back to the club page, so a DJ opens on top of the event and
    /// closing them returns here — reading a lineup shouldn't cost you the
    /// event you were reading it on.
    @ViewBuilder var djPage: (FeaturedDJ) -> DJPage

    @Environment(\.dismiss) private var dismiss
    @Environment(LocaleStore.self) private var locale
    @State private var descriptionExpanded = false
    @State private var openDJ: FeaturedDJ?

    private var lineStrong: Color { Theme.fadedSand.opacity(0.34) }

    var body: some View {
        VStack(spacing: 0) {
            topBar
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    if let raw = event.image, let url = URL(string: raw) {
                        flyer(url)
                    }
                    head
                    if let copy = descriptionText { description(copy) }
                    lineupSection
                    if let venue = event.venueName, !venue.isEmpty { venueRow(venue) }
                    Color.clear.frame(height: 8)
                }
            }
            .scrollIndicators(.hidden)
        }
        .background(Theme.cream)
        .presentationDetents([.large])
        .presentationDragIndicator(.visible)
        .sheet(item: $openDJ) { djPage($0) }
    }

    // ── Chrome ────────────────────────────────────────────────────────────────

    private var topBar: some View {
        HStack {
            Text(locale.t("event.label").uppercased())
                .font(.cfMono(9.5)).kerning(1.9)
                .foregroundStyle(Theme.fadedSand)
            Spacer()
            HStack(spacing: 8) {
                ShareLink(item: shareText) {
                    Image(systemName: "square.and.arrow.up")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(Theme.stone)
                        .frame(width: 34, height: 34)
                        .background(Theme.surfaceRaised, in: .circle)
                }
                .buttonStyle(.plain)
                Button { dismiss() } label: {
                    Image(systemName: "xmark")
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(Theme.stone)
                        .frame(width: 34, height: 34)
                        .background(Theme.surfaceRaised, in: .circle)
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.horizontal, 16)
        .padding(.top, 6).padding(.bottom, 4)
    }

    private func flyer(_ url: URL) -> some View {
        CachedAsyncImage(url: url, targetWidth: 800) {
            $0.resizable().aspectRatio(contentMode: .fill)
        } placeholder: {
            Theme.imagePlaceholder
        }
        .frame(maxWidth: .infinity)
        .frame(height: 200)
        .clipped()
        .clipShape(.rect(cornerRadius: 16))
        .padding(.horizontal, 18)
        .padding(.top, 6)
    }

    // ── Head ──────────────────────────────────────────────────────────────────

    private var head: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(kicker.uppercased())
                .font(.cfMono(9.5)).kerning(1.5)
                .foregroundStyle(Theme.ember)
                .padding(.bottom, 8)

            Text(event.title)
                .font(.cfSerif(27, italic: true))
                .foregroundStyle(Theme.ink)
                .fixedSize(horizontal: false, vertical: true)

            metaRow.padding(.top, 12)

            if let promoters = event.promoters, !promoters.isEmpty {
                Text(String(format: locale.t("detail.presentedBy"),
                            promoters.joined(separator: ", ")))
                    .font(.cfSans(12.5))
                    .foregroundStyle(Theme.fadedSand)
                    .padding(.top, 8)
                    .fixedSize(horizontal: false, vertical: true)
            }

            // Only rendered when the source actually gave a usable value — see
            // entryLabel / capacityLabel.
            if event.entryLabel != nil || event.capacityLabel != nil {
                FlowLayout(spacing: 8, lineSpacing: 8) {
                    if let entry = event.entryLabel {
                        chip(locale.t("event.entry"), entry)
                    }
                    if let cap = event.capacityLabel {
                        chip(locale.t("event.capacity"), cap)
                    }
                }
                .padding(.top, 12)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 18)
        .padding(.top, event.image == nil ? 20 : 16)
    }

    /// "Sat · Sub Rosa"
    private var kicker: String {
        [event.dateParts.weekday, event.venueName]
            .compactMap { $0 }.filter { !$0.isEmpty }
            .joined(separator: " · ")
    }

    private var metaRow: some View {
        FlowLayout(spacing: 10, lineSpacing: 8) {
            Text(dayLabel)
            if let range = event.timeRange {
                dot
                HStack(spacing: 4) {
                    Text(range.label)
                    // The end time usually belongs to the next day; say so
                    // rather than letting "23:00 – 06:00" read as one morning.
                    if range.crossesMidnight {
                        Text("+1").foregroundStyle(Theme.fadedSand)
                    }
                }
            }
            if let age = event.minimumAge {
                dot
                Text("\(age)+")
                    .padding(.horizontal, 8).padding(.vertical, 2)
                    .overlay(Capsule().stroke(lineStrong))
            }
            if let n = event.interested, n > 0 {
                dot
                Text(String(format: locale.t("event.interestedCount"), compact(n)))
            }
        }
        .font(.cfMono(10.5)).kerning(0.8)
        .foregroundStyle(Theme.stone)
    }

    private var dot: some View {
        Circle().fill(Theme.fadedSand).frame(width: 3, height: 3)
    }

    private func chip(_ label: String, _ value: String) -> some View {
        HStack(spacing: 5) {
            Text(label.uppercased())
                .font(.cfMono(9.5)).kerning(0.8)
                .foregroundStyle(Theme.stone)
            Text(value)
                .font(.cfMono(9.5, weight: .semibold)).kerning(0.8)
                .foregroundStyle(Theme.ink)
        }
        .padding(.horizontal, 10).padding(.vertical, 5)
        .background(Theme.surfaceRaised, in: .capsule)
        .overlay(Capsule().stroke(lineStrong))
    }

    // ── Description ───────────────────────────────────────────────────────────

    /// Promoter copy, but only when there is enough of it to be worth a section.
    /// A handful of rows carry a few characters of nothing.
    private var descriptionText: String? {
        guard let d = event.description?.trimmingCharacters(in: .whitespacesAndNewlines),
              d.count >= 40 else { return nil }
        return d
    }

    private func description(_ copy: String) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(copy)
                .font(.cfSans(13.5))
                .foregroundStyle(Theme.stone)
                .lineSpacing(4)
                .lineLimit(descriptionExpanded ? nil : 5)
                .fixedSize(horizontal: false, vertical: true)
            // Median copy is ~640 characters of marketing prose, so it is
            // clamped rather than allowed to push the lineup off the screen.
            if copy.count > 260 {
                Button {
                    withAnimation(.easeInOut(duration: 0.18)) { descriptionExpanded.toggle() }
                } label: {
                    Text(locale.t(descriptionExpanded ? "event.less" : "event.more").uppercased())
                        .font(.cfMono(9.5)).kerning(1)
                        .foregroundStyle(Theme.ember)
                }
                .buttonStyle(.plain)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 18)
        .padding(.top, 18)
    }

    // ── Lineup ────────────────────────────────────────────────────────────────

    private var lineupSection: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(locale.t("event.lineup").uppercased())
                .font(.cfMono(9.5)).kerning(1.9)
                .foregroundStyle(Theme.ember)
                .padding(.bottom, 12)

            if event.credits.isEmpty {
                // 29% of events bill nobody. Say that plainly instead of
                // rendering an empty section.
                Text(locale.t("event.noLineup"))
                    .font(.cfSans(13))
                    .foregroundStyle(Theme.fadedSand)
                    .fixedSize(horizontal: false, vertical: true)
            } else {
                VStack(spacing: 0) {
                    ForEach(Array(event.credits.enumerated()), id: \.element.key) { i, credit in
                        lineupRow(credit, isFirst: i == 0, showsDivider: i > 0)
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 18)
        .padding(.top, 20)
    }

    private func lineupRow(_ credit: LineupCredit, isFirst: Bool, showsDivider: Bool) -> some View {
        let dj = djFor(credit)
        return VStack(spacing: 0) {
            if showsDivider { Rectangle().fill(Theme.hairline).frame(height: 1) }
            Button {
                if let dj { Haptics.tap(); openDJ = dj }
            } label: {
                HStack(spacing: 12) {
                    avatar(dj, name: credit.name)
                    VStack(alignment: .leading, spacing: 3) {
                        Text(credit.name)
                            .font(.cfSans(14.5, weight: .semibold))
                            .foregroundStyle(dj == nil ? Theme.stone : Theme.ink)
                            .lineLimit(1)
                        // RA bills in order, so the first name is the headliner.
                        if isFirst, event.credits.count > 1 {
                            Text(locale.t("event.headliner").uppercased())
                                .font(.cfMono(9)).kerning(1)
                                .foregroundStyle(Theme.fadedSand)
                        }
                    }
                    Spacer(minLength: 8)
                    if dj != nil {
                        Image(systemName: "chevron.right")
                            .font(.system(size: 11, weight: .semibold))
                            .foregroundStyle(Theme.ember)
                    }
                }
                .padding(.vertical, 11)
                // The row is one target. Without this the Spacer between the
                // name and the chevron is empty space that hit-tests to
                // nothing, leaving two live islands with a dead gap between
                // them — and the gap is the widest part of the row.
                .contentShape(Rectangle())
                // A name we hold no page for is real and stays listed — it just
                // cannot be opened, and is dimmed so the affordance is honest.
                .opacity(dj == nil ? 0.55 : 1)
            }
            .buttonStyle(.plain)
            .disabled(dj == nil)
        }
    }

    private func avatar(_ dj: FeaturedDJ?, name: String) -> some View {
        Group {
            if let raw = dj?.imageUrl, let url = URL(string: raw) {
                CachedAsyncImage(url: url, targetWidth: 120) {
                    $0.resizable().aspectRatio(contentMode: .fill)
                } placeholder: { initialTile(name) }
            } else {
                initialTile(name)
            }
        }
        .frame(width: 40, height: 40)
        .clipShape(.circle)
    }

    private func initialTile(_ name: String) -> some View {
        Theme.surfaceRaised.overlay(
            Text(String(name.prefix(1)).uppercased())
                .font(.cfSerif(16, italic: true))
                .foregroundStyle(Theme.fadedSand)
        )
    }

    // ── Venue ─────────────────────────────────────────────────────────────────

    private func venueRow(_ venue: String) -> some View {
        HStack(spacing: 12) {
            VStack(alignment: .leading, spacing: 6) {
                Text(locale.t("event.venue").uppercased())
                    .font(.cfMono(9)).kerning(1.6)
                    .foregroundStyle(Theme.fadedSand)
                Text(venue)
                    .font(.cfSans(15.5, weight: .semibold))
                    .foregroundStyle(Theme.ink)
                    .lineLimit(2)
            }
            Spacer(minLength: 8)
        }
        .padding(.horizontal, 16).padding(.vertical, 14)
        .background(Theme.surface, in: .rect(cornerRadius: 16))
        .overlay(RoundedRectangle(cornerRadius: 16).stroke(lineStrong))
        .padding(.horizontal, 18)
        .padding(.top, 16)
    }

    // ── Formatting ────────────────────────────────────────────────────────────

    private var dayLabel: String {
        let p = event.dateParts
        return "\(p.day) \(p.month.capitalized)"
    }

    /// 1042 → "1k". The counts run from single digits to four figures.
    private func compact(_ n: Int) -> String {
        n >= 1000 ? String(format: "%.1fk", Double(n) / 1000) : "\(n)"
    }

    private var shareText: String {
        [event.title, dayLabel, event.venueName]
            .compactMap { $0 }.filter { !$0.isEmpty }
            .joined(separator: " · ")
    }
}

File: android-native/app/src/main/java/com/clubfuoco/app/features/clubdetail/EventDetailSheet.kt
package com.clubfuoco.app.features.clubdetail

import android.content.Context
import android.content.Intent
import androidx.activity.compose.BackHandler
import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Share
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.clubfuoco.app.R
import com.clubfuoco.app.core.designsystem.FuocoImage
import com.clubfuoco.app.core.designsystem.Geist
import com.clubfuoco.app.core.designsystem.GeistMono
import com.clubfuoco.app.core.designsystem.InstrumentSerif
import com.clubfuoco.app.core.designsystem.Theme
import com.clubfuoco.app.core.designsystem.clickableUnlessBusy
import com.clubfuoco.app.models.ClubEvent
import com.clubfuoco.app.models.FeaturedDJ
import com.clubfuoco.app.models.LineupCredit
import java.util.Locale

/**
 * Everything we hold about one event, opened by tapping its card on a club page.
 *
 * Drawn from measured coverage of the real data, which means it has to survive
 * four states rather than one happy path:
 *
 *   1. rich       — flyer, copy, several linked DJs, age, capacity
 *   2. no lineup  — 29% of events bill no artists at all
 *   3. floor      — no flyer AND no copy: title, time, venue, promoter only
 *   4. unlinked   — credits that are real names but not DJs we hold
 *
 * NOTHING HERE IS FAKED WHEN ABSENT. A missing flyer removes the image, a
 * missing lineup says so, and the two fields that lie — `cost` and
 * `venue_capacity` — are gated behind validators on [ClubEvent] rather than
 * shown raw.
 *
 * Port of `EventDetailSheet`.
 */
@Composable
fun EventDetailSheet(
    event: ClubEvent,
    djFor: (LineupCredit) -> FeaturedDJ?,
    onOpenDj: (FeaturedDJ) -> Unit,
    onClose: () -> Unit,
) {
    val context = LocalContext.current
    var expanded by remember { mutableStateOf(false) }
    val lineStrong = Theme.fadedSand.copy(alpha = 0.34f)
    val (weekday, day, month) = event.dateParts
    val dayLabel = "$day ${month.lowercase().replaceFirstChar { it.uppercase() }}"

    BackHandler(onBack = onClose)

    Column(
        Modifier
            .fillMaxSize()
            .background(Theme.cream)
            .safeDrawingPadding(),
    ) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                stringResource(R.string.event_label).uppercase(),
                fontFamily = GeistMono, fontSize = 9.5.sp, letterSpacing = 1.9.sp,
                color = Theme.fadedSand,
                modifier = Modifier.weight(1f),
            )
            RoundChip(Icons.Filled.Share) {
                shareEvent(context, event, dayLabel)
            }
            Spacer(Modifier.size(8.dp))
            RoundChip(Icons.Filled.Close, onClose)
        }

        Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState())) {
            event.image?.let { flyer ->
                FuocoImage(
                    flyer,
                    Modifier
                        .padding(horizontal = 18.dp)
                        .padding(top = 6.dp)
                        .fillMaxWidth()
                        .height(200.dp)
                        .clip(RoundedCornerShape(16.dp)),
                    targetWidth = 800.dp,
                )
            }

            // ── Head ─────────────────────────────────────────────────────────
            Column(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 18.dp)
                    .padding(top = if (event.image == null) 20.dp else 16.dp),
            ) {
                Text(
                    listOf(weekday, event.venueName.orEmpty())
                        .filter { it.isNotEmpty() }
                        .joinToString(" · ")
                        .uppercase(),
                    fontFamily = GeistMono, fontSize = 9.5.sp, letterSpacing = 1.5.sp,
                    color = Theme.accent,
                    modifier = Modifier.padding(bottom = 8.dp),
                )
                Text(
                    event.title,
                    fontFamily = InstrumentSerif, fontStyle = FontStyle.Italic,
                    fontSize = 27.sp, lineHeight = 31.sp, color = Theme.ink,
                )

                // Meta row — day, hours, age, interest — each part dropped when
                // absent so it never renders a stray separator.
                FlowRow(
                    Modifier.padding(top = 12.dp),
                    horizontalArrangement = Arrangement.spacedBy(10.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    Meta(dayLabel)
                    event.timeRange?.let { (label, crosses) ->
                        Dot()
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                        ) {
                            Meta(label)
                            // The end time usually belongs to the next day; say
                            // so rather than letting "23:00 – 06:00" read as one
                            // morning.
                            if (crosses) {
                                Text(
                                    "+1",
                                    fontFamily = GeistMono, fontSize = 10.5.sp,
                                    letterSpacing = 0.8.sp, color = Theme.fadedSand,
                                )
                            }
                        }
                    }
                    event.minimumAge?.let { age ->
                        Dot()
                        Text(
                            "$age+",
                            fontFamily = GeistMono, fontSize = 10.5.sp, letterSpacing = 0.8.sp,
                            color = Theme.stone,
                            modifier = Modifier
                                .clip(CircleShape)
                                .border(1.dp, lineStrong, CircleShape)
                                .padding(horizontal = 8.dp, vertical = 2.dp),
                        )
                    }
                    event.interested?.takeIf { it > 0 }?.let { n ->
                        Dot()
                        Meta(stringResource(R.string.event_interestedCount, compact(n)))
                    }
                }

                event.promoters?.takeIf { it.isNotEmpty() }?.let { promoters ->
                    Text(
                        stringResource(R.string.detail_presentedBy, promoters.joinToString(", ")),
                        fontFamily = Geist, fontSize = 12.5.sp, color = Theme.fadedSand,
                        modifier = Modifier.padding(top = 8.dp),
                    )
                }

                // Only rendered when the source actually gave a usable value.
                if (event.entryLabel != null || event.capacityLabel != null) {
                    FlowRow(
                        Modifier.padding(top = 12.dp),
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        event.entryLabel?.let {
                            FactChip(stringResource(R.string.event_entry), it, lineStrong)
                        }
                        event.capacityLabel?.let {
                            FactChip(stringResource(R.string.event_capacity), it, lineStrong)
                        }
                    }
                }
            }

            // ── Description ──────────────────────────────────────────────────
            // Only when there is enough of it to be worth a section: a handful
            // of rows carry a few characters of nothing.
            event.description?.trim()?.takeIf { it.length >= 40 }?.let { copy ->
                Column(
                    Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 18.dp)
                        .padding(top = 18.dp)
                        .animateContentSize(),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    Text(
                        copy,
                        fontFamily = Geist, fontSize = 13.5.sp, lineHeight = 19.sp,
                        color = Theme.stone,
                        maxLines = if (expanded) Int.MAX_VALUE else 5,
                        overflow = TextOverflow.Ellipsis,
                    )
                    // Median copy is ~640 characters of marketing prose, so it
                    // is clamped rather than allowed to push the lineup off the
                    // screen entirely.
                    if (copy.length > 260) {
                        Text(
                            stringResource(
                                if (expanded) R.string.event_less else R.string.event_more,
                            ).uppercase(),
                            fontFamily = GeistMono, fontSize = 9.5.sp, letterSpacing = 1.sp,
                            color = Theme.accent,
                            modifier = Modifier.clickableUnlessBusy { expanded = !expanded },
                        )
                    }
                }
            }

            // ── Lineup ───────────────────────────────────────────────────────
            Column(
                Modifier.fillMaxWidth().padding(horizontal = 18.dp).padding(top = 20.dp),
            ) {
                Text(
                    stringResource(R.string.event_lineup).uppercase(),
                    fontFamily = GeistMono, fontSize = 9.5.sp, letterSpacing = 1.9.sp,
                    color = Theme.accent,
                    modifier = Modifier.padding(bottom = 12.dp),
                )

                if (event.credits.isEmpty()) {
                    // 29% of events bill nobody. Say that plainly instead of
                    // rendering an empty section.
                    Text(
                        stringResource(R.string.event_noLineup),
                        fontFamily = Geist, fontSize = 13.sp, color = Theme.fadedSand,
                    )
                } else {
                    event.credits.forEachIndexed { index, credit ->
                        if (index > 0) {
                            Box(Modifier.fillMaxWidth().height(1.dp).background(Theme.hairline))
                        }
                        LineupRow(
                            credit = credit,
                            dj = djFor(credit),
                            isHeadliner = index == 0 && event.credits.size > 1,
                            onOpen = onOpenDj,
                        )
                    }
                }
            }

            // ── Venue ────────────────────────────────────────────────────────
            event.venueName?.takeIf { it.isNotEmpty() }?.let { venue ->
                Column(
                    Modifier
                        .padding(horizontal = 18.dp)
                        .padding(top = 16.dp)
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(16.dp))
                        .background(Theme.surface)
                        .border(1.dp, lineStrong, RoundedCornerShape(16.dp))
                        .padding(horizontal = 16.dp, vertical = 14.dp),
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(
                        stringResource(R.string.event_venue).uppercase(),
                        fontFamily = GeistMono, fontSize = 9.sp, letterSpacing = 1.6.sp,
                        color = Theme.fadedSand,
                    )
                    Text(
                        venue,
                        fontFamily = Geist, fontWeight = FontWeight.SemiBold, fontSize = 15.5.sp,
                        color = Theme.ink,
                        maxLines = 2, overflow = TextOverflow.Ellipsis,
                    )
                }
            }

            Spacer(Modifier.height(28.dp))
        }
    }
}

@Composable
private fun LineupRow(
    credit: LineupCredit,
    dj: FeaturedDJ?,
    isHeadliner: Boolean,
    onOpen: (FeaturedDJ) -> Unit,
) {
    Row(
        Modifier
            .fillMaxWidth()
            // One target for the whole row. Without this the gap between the
            // name and the chevron hit-tests to nothing, leaving two live
            // islands with a dead strip between them — and that strip is the
            // widest part of the row.
            .then(
                if (dj != null) Modifier.clickableUnlessBusy { onOpen(dj) } else Modifier,
            )
            // A name we hold no page for is real and stays listed — it just
            // cannot be opened, and is dimmed so the affordance is honest.
            .alpha(if (dj == null) 0.55f else 1f)
            .padding(vertical = 11.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Box(
            Modifier.size(40.dp).clip(CircleShape).background(Theme.surface),
            contentAlignment = Alignment.Center,
        ) {
            if (dj?.imageUrl != null) {
                FuocoImage(dj.imageUrl, Modifier.fillMaxSize(), targetWidth = 40.dp)
            } else {
                Text(
                    credit.name.take(1).uppercase(),
                    fontFamily = InstrumentSerif, fontStyle = FontStyle.Italic,
                    fontSize = 16.sp, color = Theme.fadedSand,
                )
            }
        }

        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
            Text(
                credit.name,
                fontFamily = Geist, fontWeight = FontWeight.SemiBold, fontSize = 14.5.sp,
                color = if (dj == null) Theme.stone else Theme.ink,
                maxLines = 1, overflow = TextOverflow.Ellipsis,
            )
            // The source bills in order, so the first name is the headliner.
            if (isHeadliner) {
                Text(
                    stringResource(R.string.event_headliner).uppercase(),
                    fontFamily = GeistMono, fontSize = 9.sp, letterSpacing = 1.sp,
                    color = Theme.fadedSand,
                )
            }
        }

        if (dj != null) {
            Icon(
                Icons.AutoMirrored.Filled.KeyboardArrowRight,
                contentDescription = null,
                tint = Theme.accent,
                modifier = Modifier.size(14.dp),
            )
        }
    }
}

@Composable
private fun Meta(text: String) {
    Text(
        text,
        fontFamily = GeistMono, fontSize = 10.5.sp, letterSpacing = 0.8.sp,
        color = Theme.stone,
    )
}

@Composable
private fun Dot() {
    Box(Modifier.size(3.dp).clip(CircleShape).background(Theme.fadedSand))
}

@Composable
private fun FactChip(label: String, value: String, line: androidx.compose.ui.graphics.Color) {
    Row(
        Modifier
            .clip(CircleShape)
            .background(Theme.surface)
            .border(1.dp, line, CircleShape)
            .padding(horizontal = 10.dp, vertical = 5.dp),
        horizontalArrangement = Arrangement.spacedBy(5.dp),
    ) {
        Text(
            label.uppercase(),
            fontFamily = GeistMono, fontSize = 9.5.sp, letterSpacing = 0.8.sp,
            color = Theme.stone,
        )
        Text(
            value,
            fontFamily = GeistMono, fontWeight = FontWeight.SemiBold, fontSize = 9.5.sp,
            letterSpacing = 0.8.sp, color = Theme.ink,
        )
    }
}

@Composable
private fun RoundChip(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    onClick: () -> Unit,
) {
    Box(
        Modifier
            .size(34.dp)
            .clip(CircleShape)
            .background(Theme.surface)
            .clickableUnlessBusy(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(icon, contentDescription = null, tint = Theme.stone, modifier = Modifier.size(14.dp))
    }
}

/** 1042 → "1.0k". The counts run from single digits to four figures. */
private fun compact(n: Int): String =
    if (n >= 1000) String.format(Locale.getDefault(), "%.1fk", n / 1000.0) else n.toString()

private fun shareEvent(context: Context, event: ClubEvent, dayLabel: String) {
    val text = listOf(event.title, dayLabel, event.venueName.orEmpty())
        .filter { it.isNotEmpty() }
        .joinToString(" · ")
    val intent = Intent(Intent.ACTION_SEND).apply {
        type = "text/plain"
        putExtra(Intent.EXTRA_TEXT, text)
    }
    context.startActivity(Intent.createChooser(intent, event.title))
}
